include/pybind11/detail/type_caster_base.h

```diff
@@ -31,47 +31,52 @@ PYBIND11_NAMESPACE_BEGIN(detail)
 /// A life support system for temporary objects created by `type_caster::load()`.
 /// Adding a patient will keep it alive up until the enclosing function returns.
 class loader_life_support {
+private:
+    loader_life_support* parent = nullptr;
+    std::unordered_set<PyObject *> keep_alive;
+
+    static loader_life_support** get_stack_pp() {
+#if defined(WITH_THREAD)
+        thread_local static loader_life_support* per_thread_stack = nullptr;
```

Comment:
It occurred to me that an unfortunate effect of using C++ thread_local is that every single pybind11 extension module will add an additional TLS variable --- that could make the thread-local storage rather large, as there can easily be a large number of extension modules.

To avoid that, the Python TLS API could be used instead --- the key that is allocated would have to be stored in the internals struct (which would make the ABI incompatible), or if we really want to maintain ABI compatibility, could be accessed via a separate PyCapsule that is handled in the same way as the existing internals struct.  Probably it would make more sense to just break the ABI, and take the opportunity to merge in the other ABI-breaking changes, though I don't have sufficient context to really offer much judgement on that decision.

Reply:
And why does that concern you? I suspect that many extensions may already use thread_local variables without this concern.

The problem with using the python TLS API is that without versioning the internal data structure we're left with the same issue.

Reply:
Some extensions may already use TLS, but I expect that is only a tiny fraction of them, and presumably it is providing useful functionality.

This will add an extra separate TLS variable for every pybind11 extension module (due to the use of -fvisibility=hidden, they will not be merged).  As you can easily have hundreds of extension modules loaded in a program, this is effectively adding potentially a very large number of TLS variables, and the memory usage scales with `number_of_threads_in_program * number_of_pybind11_extensions`.  Furthermore, with https://github.com/pybind/pybind11/pull/3257 taking care of the `string_view` use case, the main remaining use case is  `pybind11::implicitly_convertible`, which I think is a relatively niche feature.

Yes, the Python TLS API only helps if we ensure there is a single key shared by all pybind11 extensions.  As far as versioning the `internals` data structure, note that pybind11 already has such a mechanism, it is just a matter of bumping the abi version number --- the cost is that the type registry is not shared between different ABI versions, which may create an incompatibility between extensions compiled.  The internals data is currently accessed via a PyCapsule set as an attribute of the builtins module in the Python interpreter.  To avoid breaking the ABI, it would be possible to create a separate PyCapsule using the same mechanism that holds any additional data, in this case just the single tls key.

Reply:
You're talking about *potentially* hundreds of instances, which is admittedly an outlier.

As it stands now, I'd rather that we get this in to fix the threading issue, which I see as significant, and then we can rework it with an API version revision to use a shared python TLS variable or a pycapsule.

I'd prefer broader consensus w.r.t. the pycapsule encapsulation, though.

Reply:
I don't think hundreds is an outlier.  For example, tensorflow alone seems to have ~76.

Reply:
After we finish the PyCLIF-pybind11 integration, "hundreds" is probably very common.
But I really like this plan:
> As it stands now, I'd rather that we get this in to fix the threading issue, which I see as significant, and then we can rework it with an API version revision to use a shared python TLS variable or a pycapsule.

Reply:
Another issue that occurred to me is that if you call `py::cast` from translation unit `a.cc`, and that call is nested within an invocation of a pybind11-bound function defined in translation unit `b.cc`, and these two translation units have separate copies of pybind11 due to `-fvisibility=hidden`, then previously this would have worked (albeit not thread safe) but now it will fail.

For example, within Google's code base, this could occur if we define some utility function in a pybind_cc_library target that calls `py::cast`, then use that utility function from within a pybind11-bound function defined in a pybind_extension.  All of the pybind_cc_library targets will share a single copy of the thread_local, while each pybind_extension will have its own copy of the thread_local.

Reply:
If in the implementation of such a call you call `pybind11::cast` in such a way that a loader_life_support temporary is necessary then you would need to add a scoped instance of `loader_life_support` in the `-fvisibility=hidden` method implementation. I think that this is a really rare edge case, and as I said, I think that the state of the world here is vastly better than before, and we can change the implementation with an internal API bump.

Reply:
I noticed that there is already a `shared_data` member of the internals struct that could be used to store the TLS key without breaking ABI --- a macro could control whether to use that or a regular data member of the internals struct.  For efficiency the TLS key could also be copied to the local_internals struct.  Then when ABI is bumped the mechanism using shared_data could be removed.

Reply:
Just my two cents - I would say that this is a critical bug, of which the impact is far worse than any of the aforementioned side effects/edge cases. The team I work with have taken a fork of pybind11 and applied this fix so that we can move forward - as we have a strong requirement to run IO-bound and computationally expensive functions in parallel. Without this fix, we cannot reliably do so. I would second @laramiel's suggestion to get this fix in, and then rework/optimise later if required.

Reply:
I don't have any objection to merging this as is, especially given that there appears to be a relatively easy path to revising it to avoid duplicating the TLS variables without introducing an ABI break.